### DISCUSSION TITLE
"IndexError: string index out of range" when opening a program such as GVIM

### DISCUSSION THREAD

**Discussion by kpearsonxyz** [autokey/autokey#657] [Q&A] [answered]
I have several hotkeys that open items, such as GVIM. That script looks like:

output = system.exec_command('gvim')

So, it's simple, and I call it using a hotkey combination, WIN-G

Every time I do, however, I get an error, which I read by clicking on Tools ⇾ View Script Error, which says:

   --- Start ---
Script name: 'gVim'
Traceback (most recent call last):
 File "/usr/lib/python3/dist-packages/autokey/service.py", line 485, in execute
  exec(script.code, scope)
 File "<string>", line 1, in <module>
 File "/usr/lib/python3/dist-packages/autokey/scripting.py", line 497, in exec_command
  if output[-1] == "\n";
 IndexError: string index out of range
   --- End ---

I've looked everywhere for answers and finally gave up and just live with it. After remembering, I have a GitHub account, I decided to post this here because from what I can tell, you are good with this.

My system is Linux Mint, which is all I run, so don't know the equivalents from Windows. I've used Linux as my primary OS at work and home for ~25 years.

Thank you,
Karl

**Comment by josephj11** [comment, marked as answer]
Welcome to the AutoKey community!

Fixed in the Beta release 0.96. You are running a program that produces no output on stdout and then trying to retrieve that null output. Shouldn't cause a problem, but it does.

Either upgrade to our beta (getting help with that on our [Gitter](https://gitter.im/autokey/autokey) forum), or wrap such programs in a shell script that ends with `echo` so it returns some output to appease AutoKey.

**Reply by kpearsonxyz** [reply]
I wrote a script I call from AutoKey now instead of directly executing gvim.

No error for that script.

Now to write a bash "wrapper" so I can call one script with a command-line option, so I don't have to write a new script for everything.

That would be something like:

#!/bin/bash
#by me on 24 Jan 2022 to resolve a scripting issue with AutoKey
CMD=$1
exec $1 && echo "running $1"

That should work...

No it didn't. I had to put the echo on a separate line. Here's what the working version looks like:

  1 #!/bin/bash
  2 # from an answer on GitHub by josephj11 in lieu of upgrading for now...
  3 # Search for "IndexError: String index... there to
  4 # find my question and the answer I'm trying now.
  5 
  6 CMD=$1
  7 
  8 exec $CMD &                                                                           
  9 echo "Running $CMD"

It's working very well.

**Reply by josephj11** [reply]
Glad it works.

I have done a lot of bash and have almost never needed to use exec. It should replace the current shell/process with itself and never come back to the current script. $CMD should work by itself. You could even just run $1 without the intervening variable.

Solutions like this should be carefully evaluated and not let out into the wild because an attacker could pass anything into the script and it would run it. For personal use, it's probably not a problem.

You may need to enhance it significantly if you ever have a command file name that contains white space or other special characters or if you need to run a program that takes arguments.

Take a look at the Python subprocess module. It has several variations for calling it that can handle arguments, return codes, and retrieving results that the program writes to stdout. Passing arguments is also easier because each argument is a separate list element in the call so you don't have to assemble the whole command into one string.

**Comment by kpearsonxyz** [comment]
I never publish my scripts in an environment where someone could run them. If they want to put them in their server/workstation, they do so at their own risk. I don't give any guarantee anything I write won't cause damage to their systems, their thumbs, or their mother-in-law.

**Reply by kpearsonxyz** [reply]
Oh wait, the Python subprocess? Is autokey written in Python? Why do we use Lua?

**Reply by josephj11** [reply]
You're mixing benefits in with the risks! LOL

**Reply by josephj11** [reply]
AFAIK, there's no Lua anywhere in AutoKey. AutoKey is entirely written in Python 3.x - which is also the language our scripts are written in. All we do is add our API so they can perform keyboard, mouse, and some window oriented actions. Subprocess is a Python module. Maybe Lua has a similar facility with the same name. I have never looked into Lua.

**Reply by kpearsonxyz** [reply]
Oh, wait! Yeah, no Lua. I'm thinking of the scripts I have to write to get devilspie2 to work… Crossed my wires there.

I use devilspie2 to position the things I have auto-start when I log in, such as the position of all comm apps on the laptop display, and all service request research on another display, etc. I use xwininfo in a YAD script and take the values from that to determine everything, and the Lua part is baffling to me because it's such a different scripting language. Why did they use that and not Bash, Python, Perl, etc.?

**Reply by josephj11** [reply]
I'm a big fan of yad. I have a small bash project that uses it for its GUI interface. I'll have a look at devilspie2. New to me.

 I use xdotool for such things. I have a moderately large autostart script that compensates for some of the shortcomings of KDE's session restore. Maybe we can compare notes outside of this project.

I write a lot of bash and it does get pretty tiring trying to convince it that a string is just a string, not something that "needs" further expansion.

Join us over on [Gitter](https://gitter.im/autokey/autokey) where the developers and advanced users hang out. I have an empty room there for off-topic discussions.

(We also have a user [support forum](https://groups.google.com/forum/#!forum/autokey-users) for tier one support.)

**Reply by kpearsonxyz** [reply]
Okay, final answer, before I dive into the newer version where this has been fixed:

#!/bin/bash

CMD=$1
$CMD $2 $3
echo

And that seems to work. I tried several iterations, but in each case (through habit mostly) I was using "exec", and since that essentially starts a new instance of bash, it caused the script error, thus defeating my attempts at a solution I thought was going to work.

Now that I'm directly executing the command ($CMD) without exec, it's working. No more call last scripting error.

You told me not to use exec, or at least implied it could be my problem. It was.

And I'm now on Gitter, which I'd never heard of before it was mentioned in this thread. --Thank you!

**Reply by josephj11** [reply]
One small nit that my bash OCD won't let go of: quoting your variables tends to avoid a lot of problems.

```
#!/bin/bash
"$1" "$2" "$3" ; echo
```

See you over on Gitter!

**Comment by kpearsonxyz** [comment]
Good point. I do that many times, but forget many times, too.



---
Karl L. Pearson | ***@***.***
Pardon brevity or typos. Sent from my phone.
---

**Reply by Fazaljanan** [reply]
```
def mark_attendance(self, i, r, n):
     with open("attendance.csv", "r+", newline="\n") as f:
        myDataList = f.readlines()
        name_list = []
        for line in myDataList:
            entry = line.split((","))
            name_list.append(entry[0])
            
        if ((i in myDataList) or (r in myDataList) or (n in myDataList)):
            pass
            # speak_va("thank you your attendance has been added")
        elif ((i not in myDataList) and (r not in myDataList) and (n not in myDataList)):
            now = datetime.now()
            d1 = now.strftime("%d/%m/%Y")
            dtString = now.strftime("%H:%M:%S")
            f.writelines(f"\n{i[0]},{r},{n},{dtString},{d1},Present")
 
            conn = mysql.connector.connect(username='root', password='fazal123',host='localhost',database='face_recognizer1')
            mycursor = conn.cursor()
            if messagebox.askyesno("Confirmation","Are you sure you want to save attendance on database?"):
              for  i in myDataList:
               uid = i[0]
               uroll = i[1]
               uname = i[2]
               udate = i[3]
        if i[6] == "Present":
            utimein = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            utimeout = None
            uattend = i[6]
        else:
            utimein = None
            utimeout = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            uattend = i[6]
        qury = "INSERT INTO stdattendance(std_id, std_roll_no, std_name, std_time, std_date, std_timein, std_timeout, std_attendance) VALUES(%s,%s,%s,%s,%s,%s,%s,%s)"
        mycursor.execute(qury,(uid,uroll,uname,udate,utimein,utimeout,uattend))
        conn.commit()
        conn.close()
```


```
PS C:\Users\DELL\Documents\FaceRecog_attendaceSystem-main.final> & C:/Users/DELL/AppData/Local/Programs/Python/Python310/python.exe c:/Users/DELL/Documents/FaceRecog_attendaceSystem-main.final/Face_Recognition.py
Exception in Tkinter callback
Traceback (most recent call last):
  File "C:\Users\DELL\AppData\Local\Programs\Python\Python310\lib\tkinter\__init__.py", line 1921, in __call__
    return self.func(*args)
  File "c:\Users\DELL\Documents\FaceRecog_attendaceSystem-main.final\Face_Recognition.py", line 214, in face_recog
    img = recognize(img, clf, faceCascade)
  File "c:\Users\DELL\Documents\FaceRecog_attendaceSystem-main.final\Face_Recognition.py", line 205, in recognize
    coord = draw_boundray(img, faceCascade, 1.1, 10, (255, 255, 255), clf)
  File "c:\Users\DELL\Documents\FaceRecog_attendaceSystem-main.final\Face_Recognition.py", line 190, in draw_boundray
  File "c:\Users\DELL\Documents\FaceRecog_attendaceSystem-main.final\Face_Recognition.py", line 116, in mark_attendance
    uroll = i[1]
IndexError: string index out of range
```

**Reply by Fazaljanan** [reply]
HOW can i fix this error

**Comment by Elliria** [comment]
I think it's because both the `if i[6] == "Present":` clause and the `else` clause access the `i[6]` index, but that index doesn't exist when the `else` clause runs.

**Reply by josephj11** [reply]
@Elliria How can you tell what type of object i is when it starts out as an argument to the function? The error message seems to indicate that Python thinks it's a string.

My Python skills are minimal, but this code looks strange. The first if condition succeeds if any of i, r, or n are in myDataList, so if that fails, at least one of them is in it. So the elif can never succeed because it requires none of them to be in it.

If that's accurate and the indentation/block structure I see on the screen is also correct and not mangled somehow by GitHub, etc., there should be no way the statement `uroll = i[1]` ever gets executed, but it is shown as throwing an index error on the index of 1.

This code needs a good hard look to be sure it was coded as intended and if it gets past that, it needs to be thrown into an IDE and single stepped while checking that the intermediate values and program flow are as expected.

It appears to me that the error is a symptom of another problem rather than the problem to be solved.

EDIT: I tried one last thing. I edited the code so it's in a proper code block which fixed the initial indentation problem, but left the main problem of me misunderstanding something because as I read the code, the error that occurred is impossible because it's in a block of unintentionally dead code.

**Reply by Elliria** [reply]
The error message tells us it's a string index.

Also, the part of that code that seems inconsistent to me is the reference to `i[6]` in both clauses of the if statement even though only one of them expects it to exist, so the `else` clause is probably broken. The code hasn't gotten that far, though, because it fell over at the IndexError.

Last, but not least, I think the IndexError can be solved by accessing the list as a range, so replacing:
```
for  i in myDataList:
```
with:
```
for i in range(len(myDataList)):
```
might just do it.

**Reply by josephj11** [reply]
I love beating a dead horse. LOL

i is an input argument to the function so using it as a local index variable will destroy its original value in the function. I'm not sure of how Python handles parameters, but if it's call by reference (relatively unlikely), it will also destroy the value of i outside of the function.

However, the code reuses i in just this way even though i should be a loop invariant.

This code  appears to be written by someone just learning their first programming language who doesn't yet have the skills to keep all the moving pieces in their proper places and roles.

The exception is thrown by

`uroll = i[1]`

Since Python uses 0 based indexes, that means that i has 0 or 1 characters in it. Among other things, this implies that i does not have the contents in it that the program expects it to have. I would be wrong about this if the for loop forms it's own scope so the iterator i is not the same variable as the function argument i. If this is the case, then the variables should be renamed to avoid such confusions.

**Reply by Elliria** [reply]
I think the OP's code is expecting a number (the list index) and getting a string (the list element) instead. Since **uroll** is the second element in the OP's list, the code can't get to it. It manages to get to **uid**, which is the first element in the OP's list, because it was fed data, so it can't be empty. It can't go past that one element, though, because it wasn't given numbers. At least I think that's what's going on. I'm still very much a Python n00b.

If I'm right, then my suggested replacement line of code should fix that issue. If I'm right about the other issue, then that one will suddenly show up once the first one is fixed. Otherwise, all should be well.

Meanwhile, since it would be awfully messy for us to try to run that code up there without the necessary data, I created a couple of snippets below that can be run from a `.py` file. There are two ways that I know of so far to get at a list's indices and elements. I used `range()` above and have included it again here along with `enumerate()` as another way of doing the same thing. You can see that you've got the choice, either way, of accessing the index or the element, which you don't have with `for i in fruits:` which only gives you access to each element:
```python
# You can use enumerate to access a list's indices or elements:
print("\nENUMERATE EXAMPLE")
fruits = ["apple", "banana", "coconut"]
for i, e in enumerate(fruits):
	# print the index:
	print(i)
	# print the element:
	print(e)

# Or you can use a range in combination with the length of a list to access a list's indices or elements:
print("\nRANGE EXAMPLE")
fruits = ["apple", "banana", "coconut"]
for i in range(len(fruits)):
	# print the index:
	print(i)
	# print the element:
	print(fruits[i])
```

**Reply by josephj11** [reply]
Aside from the likelihood that this post was probably just posted in the wrong project, the elephant in the room is

```
def mark_attendance(self, i, r, n):

and 

        if ((i in myDataList) or (r in myDataList) or (n in myDataList)):
            pass
            # speak_va("thank you your attendance has been added")
        elif ((i not in myDataList) and (r not in myDataList) and (n not in myDataList)):
```

The variable i is a function parameter, so why is it being reused as a loop index? Doesn't that mess things up. If not, it's just bad programming style.

Please explain to me: how is the condition of the elif ever true?

The more I look at this code the more wrong it looks.

All the other variables have meaningful names. Why not the function parameters - which are arguably the least known because they come from a larger external scope?

`name_list` is built, but never used.

```
              for  i in myDataList:
               uid = i[0]
               uroll = i[1]
               uname = i[2]
               udate = i[3]
````

is completely fubar. On each iteration (possibly, a lot of them), those four variables keep getting overwritten by the next value without ever being used. If you just want the values from the last element in MyDataList, that's a terrible way to get them.

**Reply by Elliria** [reply]
We don't have all of the code. For example, we never actually get to see the function call that runs the main function that's holding all of that code. As far as reusing parameters as variables, that's probably considered sloppy code, but it's allowed. Check this out:
```python
print("\nEXAMPLE 1")
def example(i):

	fruits = ["apple"]

	print(i) # green

	if fruits:
		print(i) #green

	for x in fruits:
		print(x) # apple
		print(i) # green

	for i in fruits:
		print(i) # apple

	print(i) # apple

example("green")

print("\nEXAMPLE 2")
def example(i):

	fruits = ["apple"]

	print(i) # green

	if fruits:
		print(i) # green
		i = "red"

	for x in fruits:
		print(x) # apple
		print(i) # red

	if fruits:
		print(i) # red

	for i in fruits:
		print(i) # apple

	print(i) # apple

example("green")
```
You can see that in these examples, the value we gave the parameter in the function call holds the trump card until/unless another object creates a variable with the same name and gives it a different value. At that point, all other references to it will be the new value until/unless that happens again. So, I guess I'm going to have to think of Python objects as Godzilla from now on.

**Reply by Elliria** [reply]
> Please explain to me: how is the condition of the elif ever true?

We don't have the contents of MyDataList, but presumably, if what's in it doesn't match what's fed to it by this script, then the elif would be true, I suppose.

> The more I look at this code the more wrong it looks.

You'd probably have nightmares if you took a look at some of the stuff I've coded. It works, but it's not pretty. I'm certain that I can create in a room full of code what others can create in a shoe-box.

> All the other variables have meaningful names. Why not the function parameters - which are arguably the least known because they come from a larger external scope?

No idea. It could be a convention or an oversight.
 
> `name_list` is built, but never used.

I get the impression this is a work-in-progress and has hit a wall, so we're seeing it in an unfinished state.
 
> ```
>               for  i in myDataList:
>                uid = i[0]
>                uroll = i[1]
>                uname = i[2]
>                udate = i[3]
> ```
> 
> is completely fubar. On each iteration (possibly, a lot of them), those four variables keep getting overwritten by the next value without ever being used. If you just want the values from the last element in MyDataList, that's a terrible way to get them.

I totally missed this. You're absolutely right. They will always be those four values no matter how many iterations are run. I suspect it was the OP's way to try to get the indices of the list elements.

**Comment by josephj11** [comment]
Well, for starters, your console messages look like they came from Windows, so unless you're running your program under wine or proton, you're in the wrong universe because AutoKey is a Linux only application.

If you are running under wine or proton, then you may have difficulties because, while it should theoretically work, people have had problems with it and we haven't had much experience yet trying to fix them.

Your code does not appear to have any AutoKey API calls in it and you appear to be using tkinter which isn't a problem, but it's more things that makes it look like you found the wrong project to post your question.

Take a look at our [problem reporting guide](https://github.com/autokey/autokey-python2/wiki/Problem-Reporting-Guide). If you still think you're in the right place, provide the information requested there so we can figure out what your problem has to do with AutoKey and how to solve it.
